R/mash_wrapper.R

```diff
@@ -53,33 +53,40 @@ filter_invalid_summary_stat <- function(dat_list, bhat = NULL, sbhat = NULL, z =
 }
 
 #' @export
-filter_mixture_components <- function(conditions_to_keep, U, w = NULL, w_cutoff = 1e-4) {
-  # Check if colnames of Y is a subset of column names of each element in U
-  U <- lapply(U, function(mat, to_keep) {
-    missing_conditions <- setdiff(conditions_to_keep, colnames(mat))
-    if (length(missing_conditions) > 0) {
-      stop(paste("Condition(s)", paste(missing_conditions, collapse = ", "), "not found in matrix", mat_name))
-    }
-    mat[to_keep, to_keep]
-  }, conditions_to_keep)
-
-  for (mat_name in names(U)) {
-    # remove null component
-    if (all(U[[mat_name]] == 0)) {
-      U[[mat_name]] <- NULL
-      if (!is.null(w)) w <- w[!names(w) %in% mat_name]
-      next
-    }
-    if (!is.null(w)) {
-      if (w[mat_name] < w_cutoff) {
-        w <- w[!names(w) %in% mat_name]
-        U[[mat_name]] <- NULL
-      }
-      next
+filter_mixture_components <- function (conditions_to_keep, U, w = NULL, w_cutoff = 1e-04) {
+    phenos_out <- setdiff(colnames(U[[1]]), conditions_to_keep)
+    U <- lapply(U, function(mat, to_keep) {
+        missing_conditions <- setdiff(conditions_to_keep, colnames(mat))
+        if (length(missing_conditions) > 0) {
+            stop(paste("Condition(s)", paste(missing_conditions, 
+                collapse = ", "), "not found in matrix", mat_name))
+        }
+        mat[to_keep, to_keep]
+    }, conditions_to_keep)
+    
+    for (mat_name in names(U)) {
+        # remove null component
+        if (all(U[[mat_name]] == 0)) {
+            U[[mat_name]] <- NULL
+            if (!is.null(w)) 
+                w <- w[!names(w) %in% mat_name]
+            next
+        }
+        if (!is.null(w)) {
+            if (w[mat_name] < w_cutoff) {
+                w <- w[!names(w) %in% mat_name]
+                U[[mat_name]] <- NULL
+            }
+            next
+        }
     }
-  }
-  message(paste(length(U), "components of matrices remained after filtering. "))
-  return(list(U = U, w = w))
+    # remove the data driven U in U list, and recalculate w
+    U[phenos_out] <- NULL
+    w <- w[!names(w) %in% phenos_out]
+    sum_w = sum(w)
```

Comment:
sum_w has to be computed before removing any elements from `w`

Reply:
ie, before the for loop where `w` gets changed.

Reply:
I think this is related to the below comments, if we removed more context by doing your suggestion the sum of w won't be 1

Reply:
sum of w don't have to be 1. but it is important to maintain the same summed value as original in your output w. Suppose here is the input:

```
U = [U1, U2,U3]
w = [0.1, 0.2,0.1]
```

also suppose `U3` is removed, so the updated is 

```
U = [U1, U2]
```

define sum_w = 0.1 + 0.2 + 0.1 = 0.4. In this case you can understand it as the remaining 0.6 are the NULL

When `U3` is removed, the sum of the remaining weights becomes:

sum_w_new = 0.1 + 0.2 = 0.3

To maintain the original total weight sum of 0.4, adjust the weights proportionally:

w1 = (0.1 / 0.3) * 0.4 = 0.1333
w2 = (0.2 / 0.3) * 0.4 = 0.2667

So the updated weights are:

w = [0.1333, 0.2667]


My initial suggestion on slack was not correct eitehr so here i flush out it more refined fashion.

Reply:
Yes, I can switch to this version, but just a reminder: the sum_w is almost always close to 1. So, even though we normalize with (w1 / sum_w_new) * sum_w, it essentially simplifies to w1 / sum_w_new, giving us the same value as the current version. 
![image](https://github.com/user-attachments/assets/93ff2b6d-97a1-4125-af37-cd2eb4f872ec)

Reply:
yes i understand. my proposal is for when null components are also considered, as a more generic case and I can anticipate that usage.